### PR TITLE
Gossip pull response budget

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Gossip can send a huge amount of traffic

#### Summary of Changes

Have a budget for outbound requests and try to do QoS on responses.

Fixes #

### CODE REVIEW

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -101,6 +107,8 @@ pub struct ClusterInfo {
     pub(crate) keypair: Arc<Keypair>,
     /// The network entrypoint
     entrypoint: Option<ContactInfo>,
+
+    outbound_budget: (usize, u64),
```
Comment (mvines) [inline core/src/cluster_info.rs:111]: Can you please add a comment for the tuple values, it's not intuitive from the types what either value is used for

Comment (mvines) [reply core/src/cluster_info.rs:111]: Actually just make it a struct?

File: core/src/cluster_info.rs [hunk 2]
```diff
@@ -197,6 +210,12 @@ enum Protocol {
     PruneMessage(Pubkey, PruneData),
 }
 
+struct ResponseScore {
+    x: usize,
+    y: usize,
+    score: u64,
```
Comment (mvines) [inline core/src/cluster_info.rs:216]: comments for these three fields please

File: core/src/cluster_info.rs [hunk 3]
```diff
@@ -94,13 +100,20 @@ pub enum ClusterInfoError {
     BadGossipAddress,
 }
 #[derive(Clone)]
+pub struct DataBudget {
+    bytes: usize,
+    last_timestamp: u64,
```
Comment (mvines) [inline core/src/cluster_info.rs:105]: more comments here too please, it's still not obvious to me if more `bytes` is good or bad or what `last_timestamp` is used for

File: core/src/cluster_info.rs [hunk 4]
```diff
@@ -94,13 +100,21 @@ pub enum ClusterInfoError {
     BadGossipAddress,
 }
 #[derive(Clone)]
+pub struct DataBudget {
+    bytes: usize,        // amount of bytes we have in the budget to send
+    last_timestamp: u64, // Last time that we upped the bytes count,
```
Comment (mvines) [inline core/src/cluster_info.rs:105]:
```suggestion
    last_timestamp_ms: u64, // Last time that we upped the bytes count,
```
or use an `Instant`?

File: core/src/cluster_info.rs [hunk 5]
```diff
@@ -197,6 +211,12 @@ enum Protocol {
     PruneMessage(Pubkey, PruneData),
 }
 
+struct ResponseScore {
```
Comment (mvines) [inline core/src/cluster_info.rs:214]: Sorry if I'm sounding pedantic, how about details about what the x and y axis are?

**Comment by sakridge** [comment]
@mvines added those comments, hopefully a bit clearer now

File: core/src/cluster_info.rs [hunk 6]
```diff
@@ -1335,55 +1359,137 @@ impl ClusterInfo {
...
+        stakes: &HashMap<Pubkey, u64>,
     ) -> Option<Packets> {
         // split the requests into addrs and filters
         let mut caller_and_filters = vec![];
         let mut addrs = vec![];
+        let mut time = Measure::start("handle_pull_requests");
+        let now = timestamp();
+        {
+            let mut cluster_info = me.write().unwrap();
+            if now - cluster_info.outbound_budget.last_timestamp > 100 {
```
Comment (mvines) [inline core/src/cluster_info.rs:1383]: 100 -> GOSSIP_SLEEP_MILLIS?

File: core/src/cluster_info.rs [hunk 7]
```diff
@@ -1335,55 +1359,137 @@ impl ClusterInfo {
...
+                let len = std::cmp::max(stakes.len(), 2);
+                cluster_info.outbound_budget.bytes += len * 1000;
+                cluster_info.outbound_budget.bytes =
+                    std::cmp::min(cluster_info.outbound_budget.bytes, 5 * len * 1000);
```
Comment (mvines) [inline core/src/cluster_info.rs:1387]: Can we have some words or `const`s around the choice of ` * 1000` and `5 *`?

File: core/src/cluster_info.rs [hunk 8]
```diff
@@ -1335,55 +1359,137 @@ impl ClusterInfo {
...
+            let response = pull_responses[stat.x].0[stat.y].clone();
+            let protocol = Protocol::PullResponse(self_id, vec![response]);
+            packets
+                .packets
+                .push(Packet::from_data(&from_addr, protocol));
+            let len = packets.packets.len();
+            total_bytes += packets.packets[len - 1].meta.size;
+
+            if total_bytes > outbound_budget {
+                break;
```
Comment (mvines) [inline core/src/cluster_info.rs:1475]: This is where we run out of budget right?  Does this deserve a counter?

Comment (sakridge) [reply core/src/cluster_info.rs:1486]: Yep, added.

File: core/src/cluster_info.rs [hunk 9]
```diff
@@ -1335,55 +1364,148 @@ impl ClusterInfo {
...
+        let now = timestamp();
+        {
+            let mut cluster_info = me.write().unwrap();
+
+            const INTERVAL_MS: u64 = 100;
+            // allow 10kBps per staked validator, epoch slots + votes ~= 1.5kB/slot ~= 4kB/s
+            const BYTES_PER_INTERVAL: usize = 1000;
+            const MAX_BUDGET_MULTIPLE: usize = 5; // allow budget build-up to 5x the interval default
+
+            if now - cluster_info.outbound_budget.last_timestamp_ms > INTERVAL_MS {
```
Comment (mvines) [inline core/src/cluster_info.rs:1394]: `thread '<unnamed>' panicked at 'attempt to subtract with overflow', core/src/cluster_info.rs:1394:16` in CI

Comment (sakridge) [reply core/src/cluster_info.rs:1394]: Yea, that's a good catch, I need the now after I take the write lock.

**Pull request merged**